hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/om/FSORepairTool.java

```diff
@@ -0,0 +1,748 @@
+        String.class,
+        OmDirectoryInfo.class);
+    fileTable = store.getTable(OmMetadataManagerImpl.FILE_TABLE,
+        String.class,
+        OmKeyInfo.class);
+    deletedDirectoryTable = store.getTable(
+        OmMetadataManagerImpl.DELETED_DIR_TABLE,
+        String.class,
+        OmKeyInfo.class);
+    deletedTable = store.getTable(
+        OmMetadataManagerImpl.DELETED_TABLE,
+        String.class,
+        RepeatedOmKeyInfo.class);
+  }
+
+  protected static DBStore getStoreFromPath(String dbPath) throws IOException {
+    File omDBFile = new File(dbPath);
+    if (!omDBFile.exists() || !omDBFile.isDirectory()) {
+      throw new IOException(String.format("Specified OM DB instance %s does " +
+          "not exist or is not a RocksDB directory.", dbPath));
+    }
+    // Load RocksDB and tables needed.
+    return OmMetadataManagerImpl.loadDB(new OzoneConfiguration(),
+        new File(dbPath).getParentFile());
+  }
+
+  public org.apache.hadoop.ozone.repair.om.FSORepairTool.Report run() throws IOException {
+    System.out.println("Disclaimer: This tool currently does not support snapshots.");
+
+    if (bucketFilter != null && volumeFilter == null) {
+      System.out.println("--bucket flag cannot be used without specifying --volume.");
+      return null;
+    }
+
+    if (volumeFilter != null) {
+      OmVolumeArgs volumeArgs = volumeTable.get(volumeFilter);
+      if (volumeArgs == null) {
+        System.out.println("Volume '" + volumeFilter + "' does not exist.");
+        return null;
+      }
+    }
+
+    // Iterate all volumes or a specific volume if specified
+    try (TableIterator<String, ? extends Table.KeyValue<String, OmVolumeArgs>>
+             volumeIterator = volumeTable.iterator()) {
+      openReachableDB();
+
+      while (volumeIterator.hasNext()) {
+        Table.KeyValue<String, OmVolumeArgs> volumeEntry =
+            volumeIterator.next();
+        String volumeKey = volumeEntry.getKey();
+
+        if (volumeFilter != null && !volumeFilter.equals(volumeKey)) {
+          continue;
+        }
+
+        System.out.println("Processing volume: " + volumeKey);
+
+        if (bucketFilter != null) {
+          OmBucketInfo bucketInfo = bucketTable.get(volumeKey + "/" + bucketFilter);
+          if (bucketInfo == null) {
+            //Bucket does not exist in the volume
+            System.out.println("Bucket '" + bucketFilter + "' does not exist in volume '" + volumeKey + "'.");
+            return null;
+          }
+
+          if (bucketInfo.getBucketLayout() != BucketLayout.FILE_SYSTEM_OPTIMIZED) {
+            //LOG.debug("Skipping non-FSO bucket {}", bucketKey);
+            System.out.println("Skipping non-FSO bucket " + bucketFilter);
+            continue;
+          }
+
+          dropReachableTableIfExists();
+          createReachableTable();
+          markReachableObjectsInBucket(volumeEntry.getValue(), bucketInfo);
```

Comment:
it can not identify unreachable objects whose bucket is removed OR recreated. Need have that logic.

Reply:
This will be addressed in a follow-up jira.